File: Day 27 -  Hashing/Count Triangles.cpp
#include <bits/stdc++.h>
using namespace std;

int count_triangles(vector<pair<int, int>> &points)
{
  int n = points.size();
  unordered_map<int, int> freq_x;
  unordered_map<int, int> freq_y;

  for (auto P : points)
  {
    int x = P.first;
    int y = P.second;

    freq_x[x]++;
    freq_y[y]++;
  }
  int ans = 0;
  for (auto P : points)
  {
    int x = P.first;
    int y = P.second;
    int fx = freq_x[x];
    int fy = freq_y[y];

    ans += (fx - 1) * (fy - 1);
  }
  return ans;
}

int main()
{
  int n;
  cin >> n;
  vector<pair<int, int>> points(n);
  for (int i = 0; i < n; i++)
  {
    int x, y;
    cin >> x >> y;
    points[i] = {x, y};
  }
  cout << count_triangles(points);
  return 0;
}

File: Day 27 -  Hashing/Count Triangles.java
import java.util.*;

public class Main {
  static int countTriangles(List<Map.Entry<Integer, Integer>> points) {
    Map<Integer, Integer> freqX = new HashMap<>();
    Map<Integer, Integer> freqY = new HashMap<>();

    for (Map.Entry<Integer, Integer> entry : points) {
      int x = entry.getKey();
      int y = entry.getValue();

      freqX.put(x, freqX.getOrDefault(x, 0) + 1);
      freqY.put(y, freqY.getOrDefault(y, 0) + 1);
    }

    int ans = 0;
    for (Map.Entry<Integer, Integer> entry : points) {
      int x = entry.getKey();
      int y = entry.getValue();
      int fx = freqX.get(x);
      int fy = freqY.get(y);

      ans += (fx - 1) * (fy - 1);
    }
    return ans;
  }

  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    int n = scanner.nextInt();
    List<Map.Entry<Integer, Integer>> points = new ArrayList<>();
    for (int i = 0; i < n; i++) {
      int x = scanner.nextInt();
      int y = scanner.nextInt();
      points.add(new AbstractMap.SimpleEntry<>(x, y));
    }
    System.out.println(countTriangles(points));
    scanner.close();
  }
}
